{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b23e5e1c-a9c9-4577-826a-e357ea6cebbc",
   "metadata": {},
   "source": [
    "#### Author: Domitille Jarrige\n",
    "#### Date: 2022-09-21\n",
    "\n",
    "# Transposable elements mining in the _Dioszegia hungarica_ PDD-24b-2 genome\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09329a67-77f6-46a5-a6d2-d5ab87451d74",
   "metadata": {},
   "source": [
    "#### Module imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3481206c-3972-4719-baa3-8bc679b6da29",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "815f060c-be94-4136-b617-191a1985aa8b",
   "metadata": {},
   "source": [
    "## Building TE library\n",
    "\n",
    "Transposable elements (TE) were searched using protein homology to known TE from other fungi and two _de novo_ methods (RepeatModeler and EDTA).\n",
    "\n",
    "### Protein homology method\n",
    "\n",
    "BLAST+tblastx v2.11.0 was used with curated intern TE sequences (with their typical TE protein domains) from other fungi (_Saitozyma podzolica_, _Cryptococcus neoformans_, _Cryptococcus gattii_, _Rhodotorula toruloides_, _Candida glabrata_) as queries (exemple command line below)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d5b36c9-9cef-4ffb-ad7a-fe5b7aee93e7",
   "metadata": {},
   "source": [
    "`tblastx -query /path/to/transposon_queries/TE.fasta -subject /path/to/assembly/Diohu1_AssemblyScaffolds.fasta -outfmt 6 -out transposon_queries/TE_hits.txt`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b5857bf-59e5-4303-aa1e-091da75ec0a7",
   "metadata": {},
   "source": [
    "The tblastx hits were then enlarged to include 6000 bp of their flanking sequences. This was done with Biopython in a jupyter notebook as in the following exemple:\n",
    "\n",
    "##### Loading tblastx results table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef9d9b9f-d4e0-4db8-b922-00c55c72c987",
   "metadata": {},
   "outputs": [],
   "source": [
    "transposons_df = pd.read_csv(\"transposon_queries/TE_hits.tsv\", sep=\"\\t\", header=None, \n",
    "                             names=[\"transposon_name\", \"contig_name\", \"%identity\",\"alignement_length\",\"mismatches\",\"gap_open\", \"t_start\", \n",
    "                                    \"t_end\", \"diohu_start\", \"diohu_end\", \"e_value\", \"score\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722a9fec-85e8-4358-ad1d-ce0bc64fb844",
   "metadata": {},
   "source": [
    "The results tables have the following format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e98fd9e4-cb50-4678-9da1-1e6be1e42b2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transposon_name</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>%identity</th>\n",
       "      <th>alignement_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>t_start</th>\n",
       "      <th>t_end</th>\n",
       "      <th>diohu_start</th>\n",
       "      <th>diohu_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mock_TE_1</td>\n",
       "      <td>contig_16</td>\n",
       "      <td>35.146</td>\n",
       "      <td>239</td>\n",
       "      <td>155</td>\n",
       "      <td>0</td>\n",
       "      <td>1636</td>\n",
       "      <td>2352</td>\n",
       "      <td>2302085</td>\n",
       "      <td>2301369</td>\n",
       "      <td>2.210000e-92</td>\n",
       "      <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mock_TE_1</td>\n",
       "      <td>contig_16</td>\n",
       "      <td>38.967</td>\n",
       "      <td>213</td>\n",
       "      <td>130</td>\n",
       "      <td>0</td>\n",
       "      <td>220</td>\n",
       "      <td>858</td>\n",
       "      <td>2303747</td>\n",
       "      <td>2303109</td>\n",
       "      <td>2.210000e-92</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mock_TE_1</td>\n",
       "      <td>contig_16</td>\n",
       "      <td>44.737</td>\n",
       "      <td>38</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1129</td>\n",
       "      <td>1242</td>\n",
       "      <td>2302925</td>\n",
       "      <td>2302812</td>\n",
       "      <td>2.210000e-92</td>\n",
       "      <td>45.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mock_TE_1</td>\n",
       "      <td>contig_11</td>\n",
       "      <td>34.483</td>\n",
       "      <td>145</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "      <td>145</td>\n",
       "      <td>579</td>\n",
       "      <td>264963</td>\n",
       "      <td>265397</td>\n",
       "      <td>1.720000e-45</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mock_TE_1</td>\n",
       "      <td>contig_11</td>\n",
       "      <td>38.636</td>\n",
       "      <td>88</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>595</td>\n",
       "      <td>858</td>\n",
       "      <td>265422</td>\n",
       "      <td>265685</td>\n",
       "      <td>1.720000e-45</td>\n",
       "      <td>78.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transposon_name contig_name  %identity  alignement_length  mismatches  \\\n",
       "0       Mock_TE_1   contig_16     35.146                239         155   \n",
       "1       Mock_TE_1   contig_16     38.967                213         130   \n",
       "2       Mock_TE_1   contig_16     44.737                 38          21   \n",
       "3       Mock_TE_1   contig_11     34.483                145          95   \n",
       "4       Mock_TE_1   contig_11     38.636                 88          54   \n",
       "\n",
       "   gap_open  t_start  t_end  diohu_start  diohu_end       e_value  score  \n",
       "0         0     1636   2352      2302085    2301369  2.210000e-92  189.0  \n",
       "1         0      220    858      2303747    2303109  2.210000e-92  172.0  \n",
       "2         0     1129   1242      2302925    2302812  2.210000e-92   45.1  \n",
       "3         0      145    579       264963     265397  1.720000e-45  110.0  \n",
       "4         0      595    858       265422     265685  1.720000e-45   78.0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transposons_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337383a4-ffd8-40e1-8ccf-daff2f5ee3a8",
   "metadata": {},
   "source": [
    "##### Extracting TE names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32da3190-df17-435e-a45a-94f82acd617b",
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_list = list(transposons_df.groupby(\"transposon_name\").sum().index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b2fba456-e921-4852-82b5-11f0e9baeec0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Mock_TE_1', 'Mock_TE_2', 'Mock_TE_3', 'Mock_TE_4', 'Mock_TE_5', 'Mock_TE_6']\n"
     ]
    }
   ],
   "source": [
    "print(trans_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f47adf3a-d3b1-43cf-a047-17bfca881270",
   "metadata": {},
   "source": [
    "##### Extracting putative TE sequences from _D. hungarica_ PDD-24b-2 genome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f2c7d5aa-348d-4e82-9d4d-f9587fb70d32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mock_TE_1: best hit in contig_16 at pos 2302085, 2301369\n",
      "File: contig_16_putative_Mock_TE_1.fasta\n",
      "Mock_TE_2: best hit in contig_16 at pos 2302007, 2301321\n",
      "File: contig_16_putative_Mock_TE_2.fasta\n",
      "Mock_TE_3: best hit in contig_16 at pos 2302007, 2301369\n",
      "File: contig_16_putative_Mock_TE_3.fasta\n",
      "Mock_TE_4: best hit in contig_16 at pos 2302007, 2301369\n",
      "File: contig_16_putative_Mock_TE_4.fasta\n",
      "Mock_TE_5: best hit in contig_16 at pos 2302100, 2301369\n",
      "File: contig_16_putative_Mock_TE_5.fasta\n",
      "Mock_TE_6: best hit in contig_16 at pos 2302085, 2301369\n",
      "File: contig_16_putative_Mock_TE_6.fasta\n"
     ]
    }
   ],
   "source": [
    "for trans in trans_list:\n",
    "    tmp_df = transposons_df[transposons_df[\"transposon_name\"] == trans]\n",
    "    \n",
    "    # The highest scoring hit for each TE name is extracted\n",
    "    contig = \", \".join(tmp_df.loc[(tmp_df[\"score\"] == tmp_df[\"score\"].max()), \"contig_name\"].values) \n",
    "    start = \", \".join(tmp_df.loc[(tmp_df[\"score\"] == tmp_df[\"score\"].max()), \"diohu_start\"].values.astype(str))\n",
    "    end = \", \".join(tmp_df.loc[(tmp_df[\"score\"] == tmp_df[\"score\"].max()), \"diohu_end\"].values.astype(str))\n",
    "    print(f\"{trans}: best hit in {contig} at pos {start}, {end}\")\n",
    "    try:\n",
    "        start_coord = int(start.split(\", \")[0])\n",
    "        end_coord = int(end.split(\", \")[0])\n",
    "    except ValueError:\n",
    "        continue\n",
    "    parser = SeqIO.parse(\"Diohu1_AssemblyScaffolds.fasta\", format=\"fasta\")\n",
    "    for chromosome in parser:\n",
    "        if chromosome.id == contig:\n",
    "            record = SeqRecord(Seq(chromosome.seq[start_coord-6000:end_coord+6000]))\n",
    "            record.id = contig + \"putative \" + trans\n",
    "            record.description = \"putative transposon region\"\n",
    "            \n",
    "            # The files can be written but here we will just print their name\n",
    "            file_name = contig + \"_putative_\" + trans + \".fasta\"\n",
    "            print(f\"File: {file_name}\")\n",
    "            #SeqIO.write(record, file_name, format=\"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8a985ca-52c8-4fe9-99c0-51a9aa01df02",
   "metadata": {},
   "source": [
    "### RepeatModeler _de novo_ method\n",
    "\n",
    "RepeatModeler [(Flynn et al. 2020)](https://doi.org/10.1073/pnas.1921046117) is available here: https://www.repeatmasker.org/RepeatModeler/\n",
    "\n",
    "RepeatModeler v2.0.3 was used."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a4ff1ac-ede5-449d-832b-988968ee71c8",
   "metadata": {},
   "source": [
    "`BuildDatabase -name Dioszegia_hungarica Diohu1_AssemblyScaffolds.fasta`\n",
    "\n",
    "`RepeatModeler -database Dioszegia_hungarica -LTRStruct >& repeat_modeler/repeat_modeler.log`\n",
    "\n",
    "A multifasta library of putative TE was produced."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72809a36-569c-4dae-a656-e788cc7b5946",
   "metadata": {},
   "source": [
    "### EDTA _de novo_ method\n",
    "\n",
    "The Extensive de novo TE Annotator (EDTA) pipeline [(Bell et al. 2022)](https://doi.org/10.1111/1755-0998.13489) is available here: https://github.com/oushujun/EDTA\n",
    "\n",
    "`perl EDTA.pl --genome Diohu1_AssemblyScaffolds.fasta`\n",
    "\n",
    "A multifasta library of putative TE was produced."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62bec425-184b-4c3e-b1ac-a92c076e55cc",
   "metadata": {},
   "source": [
    "## Putative TE curation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a538ee3-ff95-4a7c-b0cf-8c9f165c239f",
   "metadata": {},
   "source": [
    "Protein domains typical of TE (transposase, reverse transcriptase, RNase H, integrase, aspartic protease, gag domains) were then searched in the putative sequences from the three methods with CD Search ([Marchler-Bauer and Bryant 2004](https://doi.org/10.1093/nar/gkh454); [Marchler-Bauer et al. 2017](https://doi.org/10.1093/nar/gkw1129)). CD Search is accessible here: https://www.ncbi.nlm.nih.gov/Structure/cdd/wrpsb.cgi.\n",
    "\n",
    "Terminal inverted repeats (TIR) and Target Site Duplications (TSD) were also searched to classify the TE according to [(Wicker et al. 2007)](https://doi.org/10.1038/nrg2165).\n",
    "\n",
    "**For each TE family a unique candidate with TE domains was selected for the final library (see File S1)** "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29f39e1d-4a90-4b1e-9aab-ba5f92803460",
   "metadata": {},
   "source": [
    "## TE screening with RepeatMasker\n",
    "\n",
    "Lastly the number and length of TE-related sequences (including full length and truncated copies) in the _Dioszegia hungarica_ PDD-24b-2 genome were calculated using RepeatMasker v4.1.2-p1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe3c25da-0771-4886-aa39-eb1a536fe2e7",
   "metadata": {},
   "source": [
    "`RepeatMasker -lib Supplementary_file_S1_for_RM.fasta Diohu1_AssemblyScaffolds.fasta -dir repeatmasker/ -pa 8`\n",
    "\n",
    "The RepeatMasker outpule file was then parsed, to assemble the TE fragments, with the perl tool \"One code to find them all\" [(Bailly-Bechet et al. 2014)](https://doi.org/10.1186/1759-8753-5-13), available here: http://doua.prabi.fr/software/one-code-to-find-them-all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b902347-92e9-4751-9b57-927333014834",
   "metadata": {},
   "source": [
    "`./build_dictionary.pl -rm repeat_masker/Diohu1_AssemblyScaffolds.fasta.out --unknown > repeat_masker/one_code_to_find_them_all/ltr_dictionary`\n",
    "\n",
    "`./one_code_to_find_them_all.pl --rm repeat_masker/Diohu1_AssemblyScaffolds.fasta.out --ltr repeat_masker/one_code_to_find_them_all/ltr_dictionary`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1d5ef4-4aab-480f-b5ef-b0362de48887",
   "metadata": {},
   "source": [
    "#### Table of TE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ad1458-b8f5-41fc-9619-6250af0ace00",
   "metadata": {},
   "source": [
    "##### Stats of each TE family by contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7394ff26-f5f2-4c5a-a5aa-d415f1606278",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Family</th>\n",
       "      <th>Element</th>\n",
       "      <th>Length</th>\n",
       "      <th>Fragments</th>\n",
       "      <th>Copies</th>\n",
       "      <th>Solo_LTR</th>\n",
       "      <th>Total_Bp</th>\n",
       "      <th>Cover</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>contig_11</td>\n",
       "      <td>LINE</td>\n",
       "      <td>TE-007_fam-06</td>\n",
       "      <td>2548</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>contig_11</td>\n",
       "      <td>LINE</td>\n",
       "      <td>TE-008_fam-07</td>\n",
       "      <td>2700</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3192</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>contig_11</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-11_I</td>\n",
       "      <td>5119</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>312</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>contig_11</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-15_I</td>\n",
       "      <td>6917</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8010</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>contig_11</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-16_I</td>\n",
       "      <td>247</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2284</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>contig_11</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>fam-09_I</td>\n",
       "      <td>5803</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3042</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>contig_16</td>\n",
       "      <td>DNA/TIR/Mutator</td>\n",
       "      <td>TE-001_fam-02</td>\n",
       "      <td>3082</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2647</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>contig_16</td>\n",
       "      <td>LINE</td>\n",
       "      <td>TE-008_fam-07</td>\n",
       "      <td>2700</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3336</td>\n",
       "      <td>0.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>contig_16</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-10_LTR</td>\n",
       "      <td>169</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>contig_16</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>fam-09_I</td>\n",
       "      <td>206</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5798</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>contig_19</td>\n",
       "      <td>LINE</td>\n",
       "      <td>TE-006_fam-05</td>\n",
       "      <td>5276</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1201</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>contig_19</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-11_I</td>\n",
       "      <td>194</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>504</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>contig_19</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-13_LTR</td>\n",
       "      <td>356</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>356</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>contig_19</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-15_LTR</td>\n",
       "      <td>225</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>197</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>contig_2</td>\n",
       "      <td>DNA/TIR</td>\n",
       "      <td>TE-003_fam-03</td>\n",
       "      <td>913</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>913</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>contig_2</td>\n",
       "      <td>DNA/TIR/Mutator</td>\n",
       "      <td>TE-001_fam-02</td>\n",
       "      <td>3082</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>515</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>contig_2</td>\n",
       "      <td>DNA/TIR/Mutator/MITE</td>\n",
       "      <td>TE-002_fam-02</td>\n",
       "      <td>379</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>379</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>contig_2</td>\n",
       "      <td>LINE</td>\n",
       "      <td>TE-006_fam-05</td>\n",
       "      <td>5276</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>934</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>contig_2</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-10_I</td>\n",
       "      <td>4652</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4652</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>contig_2</td>\n",
       "      <td>LTR/Copia</td>\n",
       "      <td>fam-12_I</td>\n",
       "      <td>6396</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>279</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Contig                Family        Element Length Fragments Copies  \\\n",
       "0  contig_11                  LINE  TE-007_fam-06   2548         1      1   \n",
       "1  contig_11                  LINE  TE-008_fam-07   2700         2      2   \n",
       "2  contig_11             LTR/Copia       fam-11_I   5119         2      2   \n",
       "3  contig_11             LTR/Copia       fam-15_I   6917        11      8   \n",
       "4  contig_11             LTR/Copia       fam-16_I    247         6      4   \n",
       "5  contig_11             LTR/Gypsy       fam-09_I   5803        10      9   \n",
       "0  contig_16       DNA/TIR/Mutator  TE-001_fam-02   3082         2      1   \n",
       "1  contig_16                  LINE  TE-008_fam-07   2700         3      3   \n",
       "2  contig_16             LTR/Copia     fam-10_LTR    169         2      1   \n",
       "3  contig_16             LTR/Gypsy       fam-09_I    206         3      3   \n",
       "0  contig_19                  LINE  TE-006_fam-05   5276         2      2   \n",
       "1  contig_19             LTR/Copia       fam-11_I    194         3      3   \n",
       "2  contig_19             LTR/Copia     fam-13_LTR    356         1      1   \n",
       "3  contig_19             LTR/Copia     fam-15_LTR    225         2      2   \n",
       "0   contig_2               DNA/TIR  TE-003_fam-03    913         1      1   \n",
       "1   contig_2       DNA/TIR/Mutator  TE-001_fam-02   3082         2      2   \n",
       "2   contig_2  DNA/TIR/Mutator/MITE  TE-002_fam-02    379         1      1   \n",
       "3   contig_2                  LINE  TE-006_fam-05   5276         2      2   \n",
       "4   contig_2             LTR/Copia       fam-10_I   4652         3      1   \n",
       "5   contig_2             LTR/Copia       fam-12_I   6396         1      1   \n",
       "\n",
       "   Solo_LTR Total_Bp  Cover  \n",
       "0       NaN      115   0.01  \n",
       "1       NaN     3192   0.29  \n",
       "2       0.0      312   0.03  \n",
       "3       2.0     8010   0.72  \n",
       "4       2.0     2284   0.21  \n",
       "5       5.0     3042   0.27  \n",
       "0       NaN     2647   0.11  \n",
       "1       NaN     3336   0.14  \n",
       "2       1.0      168   0.01  \n",
       "3       2.0     5798   0.25  \n",
       "0       NaN     1201   0.19  \n",
       "1       1.0      504   0.08  \n",
       "2       1.0      356   0.06  \n",
       "3       2.0      197   0.03  \n",
       "0       NaN      913   0.04  \n",
       "1       NaN      515   0.02  \n",
       "2       NaN      379   0.02  \n",
       "3       NaN      934   0.04  \n",
       "4       0.0     4652   0.21  \n",
       "5       0.0      279   0.01  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = os.scandir(\"repeat_masker/\")\n",
    "paper_transposon_df = pd.DataFrame(columns=[\"Contig\", \"Family\", \"Element\", \"Length\", \"Fragments\", \"Copies\", \"Solo_LTR\", \"Total_Bp\", \"Cover\"])\n",
    "\n",
    "contig_name = r\"(contig_[0-9]*)\"\n",
    "\n",
    "for f in files:\n",
    "    if f.name.endswith(\".copynumber.csv\"):\n",
    "        try :\n",
    "            x = re.search(contig_name, f.name).group(1)\n",
    "        except AttributeError:\n",
    "            continue\n",
    "        tmp_df = pd.read_csv(f, sep=\"\\t\", comment=\"#\")\n",
    "        tmp_df[\"Contig\"] = x\n",
    "        paper_transposon_df = pd.concat([paper_transposon_df, tmp_df])\n",
    "\n",
    "# Beginning of TE table by contig\n",
    "paper_transposon_df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "461e8db6-c553-4cc8-88c7-690716985c84",
   "metadata": {},
   "source": [
    "##### Summarized TE families statistics\n",
    "\n",
    "This table was the basis for **Table S1**. The annotation on the Class, Order, Superfamily, size, TIR, TSD and domains of the representative TE specimen from the final library (**File S1**) was added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d8c4c0e1-f409-4bcd-b48a-e02713c60ae1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fragments</th>\n",
       "      <th>Copies</th>\n",
       "      <th>Solo_LTR</th>\n",
       "      <th>Total_Bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Element</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TE-001_fam-02</th>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-002_fam-02</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-003_fam-03</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-004_fam-01</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-005_fam-04</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-006_fam-05</th>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-007_fam-06</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-008_fam-07</th>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-08_I</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-08_LTR</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-09_I</th>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-09_LTR</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-10_I</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-10_LTR</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-11_I</th>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-12_I</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-13_I</th>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-13_LTR</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-14_I</th>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "      <td>4.0</td>\n",
       "      <td>14022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-14_LTR</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-15_I</th>\n",
       "      <td>39</td>\n",
       "      <td>27</td>\n",
       "      <td>15.0</td>\n",
       "      <td>17333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-15_LTR</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8.0</td>\n",
       "      <td>571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fam-16_I</th>\n",
       "      <td>46</td>\n",
       "      <td>30</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8279</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Fragments  Copies  Solo_LTR  Total_Bp\n",
       "Element                                             \n",
       "TE-001_fam-02         10       7       0.0     10982\n",
       "TE-002_fam-02          5       5       0.0      1895\n",
       "TE-003_fam-03         10      10       0.0      7666\n",
       "TE-004_fam-01          1       1       0.0       402\n",
       "TE-005_fam-04          1       1       0.0       395\n",
       "TE-006_fam-05         19      18       0.0     12979\n",
       "TE-007_fam-06          3       2       0.0      2876\n",
       "TE-008_fam-07         14      13       0.0     12924\n",
       "fam-08_I               9       6       2.0      4408\n",
       "fam-08_LTR             1       1       1.0        97\n",
       "fam-09_I              15      14       8.0     10146\n",
       "fam-09_LTR             2       2       2.0       306\n",
       "fam-10_I               3       1       0.0      4652\n",
       "fam-10_LTR             2       1       1.0       168\n",
       "fam-11_I              14      12       1.0      7037\n",
       "fam-12_I               4       3       0.0      7035\n",
       "fam-13_I               6       4       0.0      5820\n",
       "fam-13_LTR             1       1       1.0       356\n",
       "fam-14_I              18      11       4.0     14022\n",
       "fam-14_LTR             7       7       7.0      2013\n",
       "fam-15_I              39      27      15.0     17333\n",
       "fam-15_LTR             8       8       8.0       571\n",
       "fam-16_I              46      30      15.0      8279"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_transposon_df[\"Fragments\"] = paper_transposon_df[\"Fragments\"].astype('int64')\n",
    "paper_transposon_df[\"Copies\"] = paper_transposon_df[\"Copies\"].astype('int64')\n",
    "paper_transposon_df[\"Total_Bp\"] = paper_transposon_df[\"Total_Bp\"].astype('int64')\n",
    "paper_transposon_df.groupby(\"Element\").sum()[[\"Fragments\", \"Copies\", \"Solo_LTR\", \"Total_Bp\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bc096da-27f0-423a-904d-cd940fcf774b",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Session information and modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "52089619-75f0-41c6-bca4-35ca1ee5edae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python implementation: CPython\n",
      "Python version       : 3.9.0\n",
      "IPython version      : 8.0.1\n",
      "\n",
      "Compiler    : MSC v.1916 64 bit (AMD64)\n",
      "OS          : Windows\n",
      "Release     : 10\n",
      "Machine     : AMD64\n",
      "Processor   : Intel64 Family 6 Model 165 Stepping 3, GenuineIntel\n",
      "CPU cores   : 12\n",
      "Architecture: 64bit\n",
      "\n",
      "\n",
      "pandas    : 1.3.3\n",
      "Bio       : 1.79\n",
      "jupyterlab: 3.3.4\n",
      "\n",
      "Watermark: 2.3.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -t -d -v -m\n",
    "print(\"\")\n",
    "%watermark -w -p pandas,Bio,jupyterlab"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:envBio]",
   "language": "python",
   "name": "conda-env-envBio-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
